### PR TITLE
Introduce a Kconfig COMPILE_TEST option for compile only tests

### PR DESCRIPTION
**Pull request by galak**
We have several compile only tests that get build on a lot of (if not all) of the boards.  However we don't always have the various drivers configured or supported on those boards so we need dummy device names to build things.  By introducing a Kconfig COMPILE_TEST option we can than have dummy default values for the device names regardless of support via Kconfig or dts for the true device names.

### CODE REVIEW

File: scripts/extract_dts_includes.py [hunk 1]
```diff
@@ -645,7 +645,14 @@ def generate_include_file(defs, args):
                     a = defs[node][prop].get(entry)
                     print_key_value(entry, a, maxtabstop)
             else:
+                ifdef_compile_test = False
+                if prop in compile_test:
+                    ifdef_compile_test = True
+                    print("#ifndef CONFIG_COMPILE_TEST")
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: Why print instead of sys.stdout.write, as used elsewhere in this function?

File: tests/net/all/prj.conf [hunk 2]
```diff
@@ -106,6 +107,7 @@ CONFIG_NET_CONTEXT_SYNC_RECV=y
 CONFIG_NET_CONTEXT_CHECK=y
 
 # SLIP
+CONFIG_NET_SLIP_TAP=y
```
Comment (nashif) [inline tests/net/all/prj.conf]: why do you need this? Now it is set automatically for those boards that need it, no need to add it here anymore.

File: kernel/Kconfig [hunk 3]
```diff
@@ -231,6 +231,15 @@ config BUILD_TIMESTAMP
 	help
 	Build timestamp and add it to the boot banner.
 
+config COMPILE_TEST
```
Comment (nashif) [inline kernel/Kconfig]: does not really belong under kernel/

File: Kconfig.defconfig [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+config FXOS8700_I2C_NAME
+	default ""
+
+config MAX30101_I2C_NAME
+	default ""
+
+config FXAS21002_I2C_NAME
+	default ""
+
+endif
```
Comment (nashif) [inline Kconfig.defconfig]: would add this to tests/include/ beside test.conf and avoid having a config for COMPILE_TEST, i.e. define it in the Makefile of the test directly. This way we avoid confusing with such a file being in the top of the tree.

**Review by nashif — CHANGES_REQUESTED** [review]
I would prefer to modify those test in a way where we build board with support configurations only. Adding every possible Kconfig into one file and hoping that it would work is a wrong. Trying to solve this with such a workaround makes the problem worse. If we really have to do that, I would contain all of this under tests and would not add those configurations to the main Kconfig structure. This method might also be used to hide problems instead of fixing them. So in short, I am not a big fan of this solution.

**Comment by galak** [comment]
Lets take sensors as an example, I think we are going to end up with sensor drivers that don't exist on any board.  How would we build test such a case?

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
I don't like this approach.

CONF_FILE can contain multiple config files which are merged together.
If, for a particular test case, you need all relevant drivers turned on for a particular board, create, on a per board basis additional config files which switch on all the appropriate drivers for that board.
Come up with a naming convention for it and keep them sibling to the board defconfigs (which currently don't enable all drivers by default)

**Comment by andrewboie** [comment]
Re your sensors example: I'm just confused. How would you even write such a driver without hardware to test it on?
Can we cross that bridge when we actually get there?

**Comment by nashif** [comment]
> Lets take sensors as an example, I think we are going to end up with sensor drivers that don't exist on any board. How would we build test such a case?

any sensor can be built with any board if the board supports the interfaces the sensor requires. Building every sensor for every board used to work just fine, now however something is broken and all of the sudden we are missing some Kconfigs that were removed due to DTS, probably prematurely. The catch all tests are good to a certain degree and need to be adapted and enhanced, but I would not introduce fake configuration to make this build on every board just for the sake of making this pass.

Btw, most of the sensors we have are not board specific, I think we have an issue in 1 or 2 sensors that needs to be fixed, but not using this method. For example I do not understand what happened to CONFIG_FXOS8700_I2C_NAME, this is a sensor Kconfig, how come we do not have this anymore?

**Comment by nashif** [comment]
I see what could be the actual problem here: 563f8f1464525f6b20b57a2090a4c53145a188dd

We make a sensor variable depend on the board, basically tying it to a specific board and its DTS where a sensor, even the FXOS8700 could be used on other boards, either on the board or on a shield. This Kconfig variable should not have been made DTS specific I guess. Need to dig more into all of this, but if we apply this model to all sensors we will have a mess.

**Comment by galak** [comment]
Opened Issue #1082 to track this going forward.  Going to close it out for now.

**Pull request closed without merging**